docs/MultipleLoadNativeModule.md

```diff
@@ -0,0 +1,124 @@
+    1. Generate a GUID for the client.
+    2. Create a placeholder native module and add it to hash tables.
+    3. Populate the service path (created by appending the service name to path prefix), and the module ID in the
+    native module.
+    4. Return the module HANDLE, module ID and return code `EBPF_SERVICE_NOT_PRESENT`.
+3. If runtime finds the module, and it not in "unloading" state, it will return a new handle to module.
+4. If runtime finds the module, and it is in "placeholder" state, it will block on an event, which will be set when the
+module is loaded.
+5. If runtime finds the module, and it is in "unloading" state, it will block on another event, which will be set when
+call to driver unload returns. Once this event is set, this thread will be unblocked and move to step 2.
+
+#### Misc. Changes
+1. ebpfcore will now maintain 2 hash tables to look up the native modules. One hash table has service name as key, and
+another has module ID as key.
+2. Currently native module is deleted from hash table when ref count becomes 0, which is before the driver is unloaded.
+In this new design the native module should be deleted from the hash table only when the driver unload call has
+completed. Also, we need to ensure that driver unload happens inline.
+
+### Option 2
+Another approach is to completely skip all the above synchronization logic in ebpfcore and ebpfapi.dll, and separate
+out service install / uninstall completely from the scope of the application. Native modules will be installed by some
+installer that installs the native module on the system. Once that is done, programs can be loaded / unloaded multiple
+times, without the need of install / uninstall of service entries.
+
+Changes needed to implement this:
+1. Installer installs the service entry. We need to ensure the service name is unique. Two native modules with same
+name but in different locations should be installed as 2 different services. Installer also populates the client module
+ID registry key with a random GUID.
+2. The app that loads the program will pass the service name to ebpfapi (TBD: How does app know the service name for its
+service). We can still use the existing `bpf_object__open` API, and the API can take the service name instead of file
+name as input.
+
+**Note**:
+1. One limitation with option 2 is that all existing applications that are using native eBPF programs need to now
+also have an explicit installer / un-installer for the native programs that are they are using. This installer can
+either be a separate entity (in the cases where the native eBPF program is a common program shared between multiple
+applications / products), or it can be integrated with the application that loads the program (in the cases where the
+native eBPF program is exclusively owned by that application / product).
+2. Since this approach requires multiple components to generate the service name for the sys file, one option can be to
```

Comment:
I disagree with this sentence too as noted in my comment on line 99.  How the service name can be generated based on the file path is completely orthogonal to whether to use option 1 or option 2.